engine/src/io/DataLoader.cpp

```diff
@@ -226,23 +224,42 @@ void data_loader::get_schema(Schema & schema, std::vector<std::pair<std::string,
 		schema.add_file(handle.uri.toString(true));
 	}
 
+	bool open_file = false;
+	while (this->provider->has_next()){
+		handles = this->provider->get_some(32, open_file);
+		for(auto handle : handles) {
+			schema.add_file(handle.uri.toString(true));
+		}		
+	}
+
 	for(auto extra_column : non_file_columns) {
 		schema.add_column(extra_column.first, extra_column.second, 0, false);
 	}
 	this->provider->reset();
 }
 
 std::unique_ptr<ral::frame::BlazingTable> data_loader::get_metadata(int offset) {
-	std::vector<std::shared_ptr<arrow::io::RandomAccessFile>> files;
-
-	bool firstIteration = true;
-	std::vector<data_handle> handles = this->provider->get_all();
-
-	for(auto handle : handles) {
-		files.push_back(handle.fileHandle);
+	
+	std::size_t NUM_FILES_AT_A_TIME = 32;
+	std::vector<std::unique_ptr<ral::frame::BlazingTable>> metadata_batches;
+	std::vector<ral::frame::BlazingTableView> metadata_batche_views;
+	while(this->provider->has_next()){
+		std::vector<std::shared_ptr<arrow::io::RandomAccessFile>> files;
+		std::vector<data_handle> handles = this->provider->get_some(NUM_FILES_AT_A_TIME);
+		for(auto handle : handles) {
+			files.push_back(handle.fileHandle);
+		}
+		metadata_batches.emplace_back(this->parser->get_metadata(files,  offset));
```

Comment:
what is this metadata logic that is being introduced here?

Reply:
The metadata collection takes a set of file handles and a `offset` (what is the starting file index) and gives you a metadata dataframe. Now that we can only have a set of files only at a time, we need to then collect the metadata in batches and concatenate the metadata